pyk/src/pyk/cterm/cterm.py

```diff
@@ -166,9 +166,8 @@ def match_with_constraint(self, cterm: CTerm) -> CSubst | None:
         if subst is None:
             return None
 
-        constraint = self._ml_impl(cterm.constraints, map(subst, self.constraints))  # <--
-
-        return CSubst(subst=subst, constraints=[constraint])
+        # constraint = self._ml_impl(cterm.constraints, map(subst, self.constraints))  # <--
+        return CSubst(subst=subst, constraints=cterm.constraints)
```

Comment:
Currently, `match_with_constraint` is called from the following locations (tests not included):

https://github.com/runtimeverification/k/blob/3692aa781186422363be7a49f6a157714caf84d7/pyk/src/pyk/cterm/cterm.py#L152

https://github.com/runtimeverification/k/blob/3692aa781186422363be7a49f6a157714caf84d7/pyk/src/pyk/cterm/cterm.py#L234-L235

https://github.com/runtimeverification/k/blob/3692aa781186422363be7a49f6a157714caf84d7/pyk/src/pyk/kcfg/kcfg.py#L858

https://github.com/runtimeverification/k/blob/3692aa781186422363be7a49f6a157714caf84d7/pyk/src/pyk/kcfg/kcfg.py#L937

https://github.com/runtimeverification/k/blob/3692aa781186422363be7a49f6a157714caf84d7/pyk/src/pyk/kcfg/kcfg.py#L1017

https://github.com/runtimeverification/k/blob/3692aa781186422363be7a49f6a157714caf84d7/pyk/src/pyk/kcfg/kcfg.py#L1060

Do we understand the implications of changing the behavior of `match_with_constraint` on these callers? If not, all of these call sites must be inspected to understand why `match_with_constraint` is called there. 

Alternatively, you can always introduce a new function with arbitrary behavior without breaking existing code.

Reply:
In particular, I'd like to understand the semantics of `KCFG.Cover.csubst` (as in `KCFG.create_cover` it is populated with a value from `match_with_constraint`).

@PetarMax, when doing proofs, what assumptions do you have on that attribute?

Reply:
The `Cover` is an edge from a concrete representation to a more abstract representation. `KCFG.Cover.csubst` is used to recover the source from the target. Therefore, the new implementation is correct for this use case.

Reply:
I’ve reviewed the usage you mentioned.

### CTerm.match

https://github.com/runtimeverification/k/blob/3692aa781186422363be7a49f6a157714caf84d7/pyk/src/pyk/cterm/cterm.py#L157-L158

These lines seem a bit strange to me. I don’t think `match` needs to rely on `match_with_constraint`. Implementing it this way might make things clearer.

```python
    def match(self, cterm: CTerm) -> Subst | None:
        """Find `Subst` instantiating this `CTerm` to the other, return `None` if no such `Subst` exists."""
        subst = self.config.match(cterm.config)

        if not subst or len(self.constraints) != 0 or len(cterm.constraints) != 0:
            return None

        return subst
```

I'm not sure if it is the best way to check if the CTerm is a state without constraints.

### CTerm.anti_unify

The updated `match_with_constraint` can achieve the desired functionality here, but the previous version could not. However, the check here is ineffective regardless because `match_with_constraint`'s input doesn't have a `kdef`, so it cannot perform the `implies` operation using the koreclient.

https://github.com/runtimeverification/k/blob/3692aa781186422363be7a49f6a157714caf84d7/pyk/src/pyk/cterm/cterm.py#L236-L239


### create_cover

The updated `match_with_constraint` can achieve the desired functionality here, but the previous version could not.

### split_on_constraints

The updated `match_with_constraint` can achieve the desired functionality here, but the previous version could not. Because the `Split` is an edge from souce to target, and the `CSubst` on the edge is to instantiate source to target.

### lift_split_edge

https://github.com/runtimeverification/k/blob/3692aa781186422363be7a49f6a157714caf84d7/pyk/src/pyk/kcfg/kcfg.py#L1016-L1019

The implementation here was designed to adapt to the incorrect implementation of `match_with_constraint`. We can eliminate the `.add_constraint(constraint)` when using the current, correct implementation of `match_with_constraint`.

### lift_split_split

Same as above.

Reply:
In summary, I believe these usages were intended to use `match_with_constraint` according to its comment, but the current implementation is not consistent with that intent.

https://github.com/runtimeverification/k/blob/3692aa781186422363be7a49f6a157714caf84d7/pyk/src/pyk/cterm/cterm.py#L163

Reply:
I think that here
https://github.com/runtimeverification/k/blob/d8abc6b62823435eae5a2a8df2ae1e8256ae2c43/pyk/src/pyk/kcfg/kcfg.py#L1031-L1034
I am counting on that constraint to be empty, because all those substs contain as constraints are the branching conditions.

Could `CSubst.apply` be the place where changes should be made? What if we just added the constraint there to the already existing constraints of the CTerm?

Reply:
...because it kinda does make sense for the constraint of the `CSubst` to be what it is (the implication), because it tells us what must hold for the additional constraints that the `CSubst` produces, but it doesn't quite work with the `apply`. What happens if we change the `apply`, @Stevengre?

Reply:
I'm not sure if I fully understand, but it seems the current implementation is intended to only add different constraints (`csubst.constraints`). However, the previous implementation of `CSubst.apply` didn’t actually support applying it to a `CTerm` with existing constraints.

In my current implementation, `CSubst.apply` retains only the `config` of the `CTerm`, with all constraints coming from `CSubst`.

https://github.com/runtimeverification/k/blob/c8613fd40f49f1f08ab17a83beb37db49caa2109/pyk/src/pyk/cterm/cterm.py#L333-L336

Meanwhile, `match_with_constraint` now retains only the constraints from `other`. This makes the two functions consistent, allowing them to perform their intended tasks without relying on `koreclient` operations.

For the following code, we no longer need to store constraints; we can simply use `a.cterm.match_with_constraint(cterm)`.

https://github.com/runtimeverification/k/blob/d8abc6b62823435eae5a2a8df2ae1e8256ae2c43/pyk/src/pyk/kcfg/kcfg.py#L1027-L1034

Reply:
By the way, can I go ahead and fix `anti_unification` or `anti_unify` in this PR?

Reply:
Yes, I am not sure that that implementation of `apply` is the way to go. Can we modify it just so that it adds the implication coming from the current `match_with_constraints` to the current constraints and see if that works?

Reply:
> I don’t think `match` needs to rely on `match_with_constraint`. Implementing it this way might make things clearer.

I think this change can be safely made.

> The updated `match_with_constraint` can achieve the desired functionality here

I don't think there's an agreement yet what that would be though. My main concern is, `KCFG.Cterm.csubst` is mainly populated based on the predicates the backend returns, and that might be very different to the following purely syntactic requirement:

>  `KCFG.Cover.csubst` is used to recover the source from the target.

Reply:
This is the current test for `match_with_constraints`, which I believe also tests `apply`. Is that what you want?

https://github.com/runtimeverification/k/blob/47911b5591eea6bcb01ef8296a75b416b9a6c229/pyk/src/tests/unit/test_cterm.py#L81-L97